### PR TITLE
Initial Pull Request

### PR DESCRIPTION
**Pull request by kahnvex**
This pull includes:
- Source for jquery messages
- Some basic tests using qunit
- Documentation

Still to do:
- Documentation for running tests and creating a dist
- Make this bower installable
- Switch testing to Jasmine

@geekytime I hear you are the Yola javascript guy, could you review when you have extra time?

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+  "name": "jquery-plugin",
+  "version": "0.0.0-ignored",
+  "engines": {
+    "node": ">= 0.8.0"
+  },
+  "scripts": {
+    "test": "grunt qunit"
+  },
+  "devDependencies": {
+    "grunt-contrib-jshint": "~0.6.0",
```
Comment (geekytime) [inline package.json]: I've never been burned by this personally, but I've heard a few alleged horror stories around using the `~` for versions.

I feel like [Charlie's recommendation](http://blog.nodejitsu.com/package-dependencies-done-right#best-practices) to use the "0.6.x" format is probably good advice.

File: package.json [hunk 2]
```diff
@@ -0,0 +1,19 @@
+{
```
Comment (geekytime) [inline package.json:1]: @thomasw, @blaix 
NPM and Bower both use a 2-space indent when they write/overwrite their json files. Which is probably why this one has a 2-space indent.

Is it okay if we allow 2-space indents, just for `package.json` and `bower.json`?

Otherwise we're going to have to keep spotting and correcting the indents by hand every time we add a dependency via `--save` or `--save-dev`.

File: package.json (hunk 1 above)
Comment (kahnvex) [inline package.json]: Should I simply remove the `~`?

File: Gruntfile.js [hunk 3]
```diff
@@ -0,0 +1,87 @@
+'use strict';
```
Comment (geekytime) [inline Gruntfile.js:1]: I'm not opposed to `'use strict'`, but I have no idea what we gain by having it in our Gruntfile. It looks like Ben Alman puts them in [Grunt's own Gruntfile](https://github.com/gruntjs/grunt/blob/master/Gruntfile.js), so at least we're in good company. :wink:

File: package.json (hunk 2 above)
Comment (blaix) [inline package.json:1]: I think we need to be viewing and cleaning up those files by hand before committing regardless of our indent "rules", but I'd be fine with a 2-space indent since that's what the tools seem to prefer and the files aren't edited very often.

File: Gruntfile.js [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+module.exports = function(grunt) {
+
+  // Project configuration.
```
Comment (geekytime) [inline Gruntfile.js]: Do we really need any of these line comments anywhere in the file?

Comment (kahnvex) [inline Gruntfile.js]: I think these are inserted when Grunt creates the jquery plugin boiler plate after running `grunt-init jquery` I can remove them if you'd like.

File: dist/messages.js [hunk 5]
```diff
@@ -0,0 +1,124 @@
+/*! Messages - v0.1.0 - 2013-09-19
+* https://github.com/yola/jquery-messages
+* Copyright (c) 2013 kahnjw; Licensed MIT */
+;(function($){
+    var pluginName = 'Messages',
+        defaults = {
```
Comment (geekytime) [inline dist/messages.js:4]: Do we want to provide a default value for `message_attribute`, here?

File: dist/messages.js [hunk 6]
```diff
@@ -0,0 +1,124 @@
+/*! Messages - v0.1.0 - 2013-09-19
+* https://github.com/yola/jquery-messages
+* Copyright (c) 2013 kahnjw; Licensed MIT */
+;(function($){
```
Comment (thomasw) [inline dist/messages.js]: Missing `'use strict';`

File: dist/messages.js (hunk 5 above)
Comment (kahnvex) [inline dist/messages.js:4]: Good call, I'll make this change ASAP. Going to lunch now :hamburger:

File: examples/example-1.html [hunk 7]
```diff
@@ -0,0 +1,104 @@
+<html>
```
Comment (geekytime) [inline examples/example-1.html]: Can we rename this file to `basic-example.html` or just `example.html`?

I'm not a big fan of numbers in filenames. They don't mean anything.

File: examples/example-1.html [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+            }
+            .container {
+                width: 18%;
+                float: left;
+            }
+        </style>
+        <script src="http://ajax.googleapis.com/ajax/libs/jquery/1.10.2/jquery.min.js"></script>
+        <script src="../src/messages.js"></script>
+        <script>
+            $(document).ready(function() {
```
Comment (geekytime) [inline examples/example-1.html]: Can we please move all of the script tags down to the bottom of the page, so we don't need the `$(document).ready()` anymore? Bottom-loading our scripts is always a good idea.

(I know this is just an example, but it could still _set_ a better example. Ha ha.)

File: examples/example-1.html [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+                        attach_to: 'input'
+                    });
+                    return false;
+                });
+            });
+        </script>
+    </head>
+    <body>
+        <h1>Messages Example 1</h1>
+        <div class="container-1 container">
```
Comment (geekytime) [inline examples/example-1.html]: Using more meaningful class names for these elements would make it easier to read and understand the example.

How about `success-container`, `info-container`, `error-container`, `clear-container`.

File: examples/example-1.html [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+            <form class="form-1">
+                <span>Insert success</span>
+                <input type="submit" value="Insert">
+            </form>
+        </div>
+
+        <div class="container-2 container">
+            <form class="form-2">
```
Comment (geekytime) [inline examples/example-1.html]: You could get away with much simpler markup, here:

```
<div class="info-container">
    <input type="button" value="Insert Info" />
</div>
```

Then you can just handle the `click` event in the js.

Unless there's a reason the forms are important to the example, I guess. In which case, a separate forms example with more explanation might be good.

File: dist/messages.js [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+            message.html(this.message_details);
+        }
+
+        var element = null;
+        if(this.message_details.field_name){
+            var element_selector = '[name="' + this.message_details.field_name + '"]';
+            element = $(this.element).find(element_selector);
+            message.attr('class', 'fieldError alert ' + message_class).hide();
+            element.parent().append(message);
+            message.slideDown(200);
```
Comment (geekytime) [inline dist/messages.js]: Can we please drop the jquery animations in favor of [CSS transitions](https://developer.mozilla.org/en-US/docs/Web/Guide/CSS/Using_CSS_transitions)?

The CSS transitions simplify the code, perform better, gracefully degrade in older browsers, and provide better separation of content, functionality, and presentation.

File: dist/messages.js (hunk 5 above)
Comment (geekytime) [inline dist/messages.js:4]: Okay, I get it better now. 

You can use this to provide multiple messages in the content, and then refer to them by name in the code. 

I guess a default attribute wouldn't help much, unless they only wanted to use a single message, in which case they could simply call  `$(".myElement").displayMessage()`.

File: examples/example-1.html [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+                        message: 'This error is red. Like a firetruck.',
+                        message_type: 'error',
+                        attach_to: 'input',
+                        clean: j_clean
+                    });
+                    return false;
+                });
+
+                $('.remove').submit(function() {
+                    $('.alert').remove();
```
Comment (geekytime) [inline examples/example-1.html]: Can we maybe provide a method on the plugin to have the alerts remove themselves? You can check out the bootstrap tooltip, popover, and alert for examples. The user's code could look like this:

```
//One way to close one
$(".my-element").displayMessage("close");

//Another way to close one
$.displayMessage.close(".my-element");

//Close all
$.displayMessage.closeAll();
```

File: dist/messages.js (hunk 6 above)
Comment (kahnvex) [inline dist/messages.js]: Should we really put `'use strict';` in the distribution copy?

File: src/messages.js [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+            alert_classes: {
+                'info': 'alert-info',
+                'loading': 'alert-info loading',
+                'error': 'alert-error alert-danger',
+                'success': 'alert-success',
+                'alert': 'alert'
+            }
+        };
+
+    $.fn.displayMessage = function(options){
```
Comment (geekytime) [inline src/messages.js]: I kind of like the usual convention of Jquery plugin functions being nouns instead of verbs. (e.g. in Bootstrap, it's: `tooltip`, and 'popover', not `showTooltip` or `displayPopover`.)

Can we change this to `$.fn.message`?

File: dist/messages.js (hunk 6 above)
Comment (geekytime) [inline dist/messages.js]: Do we really need 'use strict' at all, since we're linting our code?

File: dist/messages.js (hunk 11 above)
Comment (kahnvex) [inline dist/messages.js]: I'll make this change tonight.

Comment (geekytime) [inline dist/messages.js]: Before you do that...

@blaix, @thomasw, do you have any opinions on switching to CSS transitions?  IE9 would not see the animations if we do that, but everything would otherwise work. There are [polyfills for this](http://www.useragentman.com/blog/csssandpaper-a-css3-javascript-library/), but I've never bothered with them. (IE9 users deserve their sub-par web experience.)

File: dist/messages.js (hunk 6 above)
Comment (thomasw) [inline dist/messages.js]: geekytime: Yes.

kahnjw:

``` javascript
;(function($){
   'use strict';
   ...
})(jQuery);
```

File: dist/messages.js (hunk 11 above)
Comment (kahnvex) [inline dist/messages.js]: Should I just make the jquery slide down feature an option? This way if you want to use CSS3 transitions you can. I feel like CSS3 transitions should be done outside this plugin.

File: dist/messages.js [hunk 14]
```diff
@@ -0,0 +1,139 @@
+/*! Messages - v0.1.0 - 2013-09-19
+* https://github.com/yola/jquery-messages
+* Copyright (c) 2013 kahnjw; Licensed MIT */
+;(function($){
+    var pluginName = 'Messages',
+        defaults = {
+            alert_classes: {
+                'info': 'alert-info',
```
Comment (geekytime) [inline dist/messages.js]: Could we maybe simplify the whole 'type'/'class' thing to just use classes?

```
$(".my-element").message({
  "message":"An error! Oh noes!",
  "class":"error"
});
```

Then they can use the built-in messages classes, or specify their own. And they don't really need to know any additional info about how it works, either way. I think this is more how the Bootstrap JS plugins work...

File: dist/messages.js [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+        return this.message_details.message !== undefined;
+    };
+
+    Messages.prototype.cleanMessages = function(){
+        var clean = null;
+        if (this.options.clean === true || this.options.clean === false) {
+            clean = this.options.clean;
+        }
+        else { clean = true; }
+
```
Comment (geekytime) [inline dist/messages.js]: Can we normalize the `clean` value when the plugin is initialized, as we merge the defaults and options?

File: dist/messages.js (hunk 6 above)
Comment (kahnvex) [inline dist/messages.js]: Added

File: dist/messages.js (hunk 11 above)
Comment (geekytime) [inline dist/messages.js]: I think we should provide the basic styles and the css animations in two separate less files, and let the user decide which styles they want to use. We could even provide multiple animation styles this way, if we wanted to.

The important thing is to separate the presentation from the functionality, and not lock the user into a particular functionality. If we want to provide the Jquery animations as an option, it'd be cleanest if we could do it via a mixin, or some other way to keep the code a bit more separated.

Is this something that can wait for a future pull, once you get Less and Bower involved?

**Comment by geekytime** [comment]
This is looking really cool. Assuming you're switching to Jasmine and we can mess with Less-based css animations in a future pull, I just have two more questions/comments:
- I know these are just dev dependencies, and I'm glad you fixed the `~` thing, but [if we use the "0.4.x" notation](http://blog.nodejitsu.com/package-dependencies-done-right#best-practices), we'll be more likely to get bug-fixes sooner, as we'll get the minor, non-breaking version bumps automatically. This idea also applies to bower dependencies.
- I'd still be interested in seeing if we could collapse the whole type/class thing down to just use classes:

> Could we maybe simplify the whole 'type'/'class' thing to just use classes?
> 
> ```
> $(".my-element").message({
>   "message":"An error! Oh noes!",
>   "class":"error"
> });
> ```
> 
> Then they can use the built-in messages classes, or specify their own. And they don't really need to know any additional info about how it works, either way. I think this is more how the Bootstrap JS plugins work...

File: Gruntfile.js (hunk 4 above)
Comment (geekytime) [inline Gruntfile.js]: I usually get rid of these kinds of things, yeah.

**Comment by geekytime** [comment]
Oh yeah, one more thing. I haven't seen those banner comments in any of our other code. Is that something this library needs?

**Comment by kahnvex** [comment]
It took me some time to reconcile changing the `message_type` option to `class`, but I think in the end it makes sense and will make it easier for the user. I think I'll actually make it `classes` and allow the user to specify multiple classes, so they can mixin different things like loading icons or different font sizes along with the message display color and all that. 

I'll also change the requirements to `0.4.x` notation so we get bug fixes.

The banner comments in /dist/ are generated by Grunt and I don't know how to make it not print them out. I'll look into it, there has to be a way.

**Comment by geekytime** [comment]
You can just pull the `banner:` section out of the gruntfile to remove those banner comments.

**Comment by kahnvex** [comment]
@geekytime Got those changes in.

File: package.json (hunk 2 above)
Comment (thomasw) [inline package.json:1]: I'm fine with it. I'm also fine with adopting two spaces for all JS and HTML, but that is a separate discussion.

Comment (blaix) [inline package.json:1]: lets go with no spaces

File: dist/messages.js [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+    $.fn.message = function(options){
+        return this.each(function() {
+            $.data(this, 'plugin_' + pluginName,
+            new Messages(this, options));
+        });
+    };
+
+    $.message = {};
+
+    $.message.close = function(selector) {
```
Comment (geekytime) [inline dist/messages.js]: Can we make the close() function a member of the Messages class, similar to the `cleanMessages` function? Then we should be able to use `this.element` to remove the alert.

File: dist/messages.js [hunk 17]
```diff
@@ -0,0 +1,149 @@
...
+        $(selector).remove();
+    };
+
+    $.message.closeAll = function(classes) {
```
Comment (geekytime) [inline dist/messages.js]: I know I suggested this closeAll function, but I didn't really think about how much work it would be to do it right. Do we really even need this in any real life scenarios?

If not, let's just remove it. This implementation makes me nervous, because it could often end up doing things the developer wouldn't expect.

If we do need it, what would you think of keeping track of the Message instances and clearing them out by calling each instance's close method?

File: dist/messages.js (hunk 16 above)
Comment (kahnvex) [inline dist/messages.js]: Are you suggesting doing something like `$('#my_selector').message({clean: true});`?

Comment (geekytime) [inline dist/messages.js]: Well, sort of, I didn't explain that very well. I guess it seems like with most jquery plugins, you can send string commands to their 'fn.function' after they're initialized.

The best Jquery plugins I've used work as stand-alone JS classes, and then also initialize sepjQuery helper functions, to make it easier to call them via Jquery. The bootstrap [tooltip plugin initalization code](https://github.com/twbs/bootstrap/blob/master/js/tooltip.js#L364) shows a simple example of how this works. 

The first time you call $(".myElement").tooltip(), it calls the Tooltip constructor, and stores the instance in the jquery data. The next time you call the jquery function for that same element, if you pass it a string command ('hide', 'show', 'toggle', 'destroy'), it invokes the appropriate method on the Popover class instance. It delegates responsibility of just about everything to the Tooltip class.

More importantly, the selectors are never used again by the code. The instance of the Popover class has a reference to the element, and the element has a reference to the class via the data property. If we really need to be able to show/hide/destroy these alerts, this is the cleanest way to do it, and will result in fewer surprises for the consumer of the plugin. We shouldn't have to pass selectors to jquery plugins, that's what the $ method is for. And removing elements from the dom based on 'default' selectors is a little scary to me. :scream:

Does that make sense? I'm sorry if I'm making things overly complicated.

Comment (kahnvex) [inline dist/messages.js]: Yes that makes sense to me. This is probably another thing I will take care of once I am finished with my prioritized work. I'll ping you when I've made the changes.

File: dist/messages.js [hunk 18]
```diff
@@ -0,0 +1,151 @@
+;(function($){
```
Comment (imorrison) [inline dist/messages.js]: You should pass in `window, document, undefined` after the dollar sign and then invoke the function with `jQuery, window, document`. This is a best practice: http://addyosmani.com/resources/essentialjsdesignpatterns/book/#jquerypluginpatterns

File: dist/messages.js [hunk 19]
```diff
@@ -0,0 +1,151 @@
...
+    Message.prototype.close = function() {
+        $(this._message_).remove();
+    };
+
+    Message.prototype.buildMessage = function(){
+        var element_attribute = $(this.element).data(this.options.message_attribute);
+        var body_attribute = $(document.body).data(this.options.message_attribute);
+        var fallback_attribute = $(document.body).data(this.options.fallback_message_attribute);
+
+        if (this.options.message_attribute && element_attribute) {
```
Comment (imorrison) [inline dist/messages.js:73]: This might be a good place for a case/switch

Comment (kahnvex) [inline dist/messages.js:73]: This would be difficult because switch statements usually "switch" over one object, not different objects.

**Comment by geekytime** [comment]
Do we want to take the `banner` section out of the gruntfile?

**Comment by geekytime** [comment]
Assuming we can tackle Jasmine tests and CSS animations in the next pull, this is probably close enough. 

It'd be nice to break a few of those longer functions up into smaller, more descriptive pieces, but it might be wise to do that as you start building the Jasmine tests. Writing the tests will help you decide which blocks of code need to be carved out into testable chunks.

@blaix, @thomasw, any additional thoughts?

**Comment by kahnvex** [comment]
Banner section removed from gruntfile

**Comment by kahnvex** [comment]
http://wheningit.tumblr.com/post/31693267852/when-a-co-worker-ignores-my-pull-request

File: LICENSE-MIT [hunk 20]
```diff
@@ -0,0 +1,22 @@
+Copyright (c) 2013 kahnjw
```
Comment (blaix) [inline LICENSE-MIT]: copyright yola?

**Comment by kahnvex** [comment]
@geekytime @blaix @thomasw Would really like to get thumbs upped on this before we merge https://github.com/yola/yolacom/tree/login_refactor so we can be using an up to date distribution of this.

**Comment by geekytime** [comment]
Looks like a great start! :+1:
